### ISSUE TITLE
always "Opening Remote"

### ISSUE THREAD

**Issue by zkd8907** [microsoft/vscode-remote-release#546]
My server is behind the proxy and I've set the config of proxy following [Set HTTP_PROXY / HTTPS_PROXY on the remote host](https://code.visualstudio.com/docs/remote/troubleshooting). Now I can use `wget https://www.microsoft.com` at server and get the right response. When I use Visual Studio Code to connect the server, it always says "Opening Remote..." and terminal shows the information below:
```
Installing to /root/.vscode-server-insiders/bin/0284236851a94b116f468345f6e98688a737015d...
Downloading with wget
```
I cannot know which file Visual Studio Code is downloading and how to resolve it.

Here is my Visual Studio Code Version Info: Version 1.35.0-insider (1.35.0-insider)

**Comment by zkd8907** [comment]
After a long time, I received this message from "OUTPUT /  Remote - SSH":
```
remote-ssh@0.39.0
SSH Resolver called for "ssh-remote+7b22686f73744e616d65223a225859e6ada3e5bc8fe78eafe5a283227d"
SSH Resolver called for host: server-name
Setting up SSH remote "server-name"
Using commit id "0284236851a94b116f468345f6e98688a737015d" and quality "insider" for server
Install and start server if needed
> Installing to /root/.vscode-server-insiders/bin/0284236851a94b116f468345f6e98688a737015d...
> Downloading with wget
> 
> 53d1fb8e-55b3-487a-9f55-07826e90ed99##25##
"install" terminal command done
Received install output: 53d1fb8e-55b3-487a-9f55-07826e90ed99##25##
Server download failed
Downloading VS Code Server failed. Please try again later.

```

**Comment by ansiblejunky** [comment]
I have the same problem! Would like to understand how to resolve this. +1

**Comment by billychou** [comment]
I have the same problem! Would like to understand how to resolve this. +1
remote-ssh@0.41.0

**Comment by ansiblejunky** [comment]
Here is the output I get in the terminal window when attempting to connect to remote server.
```
remote-ssh@0.41.0
SSH Resolver called for "ssh-remote+myserver.net"
SSH Resolver called for host: myserver.net
Setting up SSH remote "myserver.net"
Using commit id "5c65d9bfa4c56538150d7f3066318e0db2c6151f" and quality "insider" for server
Using SSH config file "F:\.ssh\config"
Install and start server if needed
> 
> Migrating .vscode-remote to /home/U12345/.vscode-server-insiders...
> Installing to /home/U12345/.vscode-server-insiders/bin/5c65d9bfa4c56538150d7f30
> 66318e0db2c6151f...
> cp: cannot stat '/home/U12345/.vscode-remote/data': No such file or directory
> cp: cannot stat '/home/U12345/.vscode-remote/extensions': No such file or directory
> Downloading with wget
> 
> 
> 5a8da578-68a5-406e-91dc-c3210f603d19##25##
> 
"install" terminal command done
Received install output: 5a8da578-68a5-406e-91dc-c3210f603d19##25##
Server download failed
Downloading VS Code Server failed. Please try again later.
```

**Comment by ansiblejunky** [comment]
I found a workaround for this issue [here](https://github.com/microsoft/vscode-remote-release/issues/15) but the instructions are a little hard to follow imho. Here's a breakdown of what it's doing so you have a better understanding:

Basically, the issue above is that VSCode performs ssh to remote server and is attempting to download the vscode-server package using `wget`. It hands because most likely the remote server requires a proxy to get to outside world.  One work around is simply to add the proxy info to the `.bash_profile` file so that they are automatically set and VSCode can do its normal thing. 

Alternatively, you can manually workaround that.  Here's what was suggested but with more info around it all from my experience today.
manual solve, only for ssh mode

* Remote login to server
* Download vscode-server to /tmp folder
Run the following `wget` command to get the correct version of `vscode-server-linux-x64.tar.gz`. Replace the COMMIT_ID with the value that was in the terminal window when it failed. It should've told you what commit it was trying to pull. This is important because the Remote SSH extension needs a specific matching version of the vscode-server otherwise you will get a `version mismatch` error.  
Clearly you might have to do this on your Windows server and the push the file to the remote server since the remote server had no outside access (the original problem!).
```
COMMIT_ID=5c65d9bfa4c56538150d7f3066318e0db2c6151f
wget -nv -O /tmp/vscode-server-linux-x64.tar.gz https://update.code.visualstudio.com/commit:$COMMIT_ID/server-linux-x64/insider
```

* On your Windows system where VSCode Insiders is installed, modify a file in the Remote SSH extension so that it no longer performs the `wget` command and instead copies the file from the /tmp folder.
Modify `C:\Users\<user>\.vscode-insiders\extensions\ms-vscode-remote.remote-ssh-<version>\out\extension.js`.

Search for `wget -nv -O vscode-server-linux-x64.tar.gz https://update.code.visualstudio.com/commit:$COMMIT_ID/server-linux-x64/${t}` and replace it with `cp /tmp/vscode-server-linux-x64.tar.gz`.

* Cleanup the remote server by removing `~/.vscode-server-insiders`
* Re-run the "Connect to remote server" and it should work now

Good luck!

**Comment by zkd8907** [comment]
@ansiblejunky 
It seems that the proxy of my server cannot reach `https://update.code.visualstudio.com/commit`. Your solution should work, but it may invalid when the version updated?

**Comment by ansiblejunky** [comment]
@zkd8907 yep, anytime you modify the files related to the extensions or app, you will have to re-apply this each time you upgrade the extension, but at least its going to work ;) alternatively, if you can think of a good solution then do a PR on the Remote SSH extension. good luck :)

**Comment by zkd8907** [comment]
@ansiblejunky 
Thanks a lot~~ I will try it later. If I get other way to solve it, I will post it here.

**Comment by MengJie** [comment]
Yes. I have the same problem.
I check the environment variables of wget process using cat /proc/[pid]/environ. I found the http[s]_proxy variable is not set.
For workaround, I set proxy in $HOME/.wgetrc like:
http_proxy = http://proxy[:port]
https_proxy = http://proxy[:port]
the wget can download successful.
But after that, Intall extension on the remote server is still not work.

**Comment by ansiblejunky** [comment]
@MengJie Yes! We also figured out using the `~/.wgetrc` is the best solution. But as you mentioned we also are hitting the next problem which is installation of extensions! Seems its a proxy issue again but I am not sure where to fix that.

**Comment by zkd8907** [comment]
@MengJie @ansiblejunky 
I use your solution and it works for me, although extensions still cannot be installed.

**Comment by MengJie** [comment]
@zkd8907 which shell are you using?
for Bash, make sure there is NO following lines at the begin of your .bashrc, or move the proxy setting before it:
```
# If not running interactively, don't do anything
case $- in
    *i*) ;;
      *) return;;
esac
```
for zsh, set the proxy configuration in $HOME/.zshenv, and maybe PATH need set too.

You can check the environment variables using following command:
`ssh host printenv`

Hope it helps.

**Comment by ansiblejunky** [comment]
That doesn't seem to fix the problem with installing extensions. Haven't figured out a good solution for that yet.

**Comment by MengJie** [comment]
@ansiblejunky What's your result of `ssh host printenv`?

**Comment by ansiblejunky** [comment]
@MengJie I assume you mean just `printenv` command? Without posting some sensitive info on github, can you tell me what you are asking for exactly? Does it resolve the installation of extensions issue?

**Comment by MengJie** [comment]
@ansiblejunky No, not just `printenv`, using `ssh host printenv` on the pc you running vscode. check if there is `http[s]_proxy`. This is the way run `printenv` in a non-interactive shell.
Currently the vscode run remote command using non-interactive shell. It might be the shell setting in non-interactive mode that not configure proxy right.

**Comment by roblourens** [comment]
Is everyone using a proxy, and possibly missing the http[s]_proxy env vars? If so it's likely that https://github.com/microsoft/vscode-remote-release/issues/546#issuecomment-500648208 is the cause, ie the variables are set in .bashrc but after a check for an interactive shell. In that case, we should be probing the user environment in an interactive shell, same as vscode does when it starts up.

If that is not the case, please include more details like the full log from the Remote-SSH output channel. Also don't be afraid to open your own issue if you aren't sure whether yours is the same as others.

**Comment by pfebolo** [comment]
I'm behind a corporative proxy and my solution was to set up the proxy info in /etc/environment file inside of the WSL environment..

```(bash)
 sudo nano /etc/environment
```
aggregate
```(bash)
 HTTP_PROXY=http://[domain]%5C[user]:[pass]@[ip_proxy]:[port]
 HTTPS_PROXY=http://[domain]%5C[user]:[pass]@[ip_proxy]:[port]
 http_proxy=http://[domain]%5C[user]:[pass]@[ip_proxy]:[port]
 https_proxy=http://[domain]%5C[user]:[pass]@[ip_proxy]:[port]
```
and save
> Notes: 
> * (1): I used IP address, I didn't use DNS name 
> * (2): The %5C is the backslash  character '\\' but encodedfy with URL encoded.

**Comment by cooperkuo** [comment]
you can check out your sshd log. In my case, when I turn **AllowTcpForwarding** to yes, vscode can remote to my server.

**Comment by roblourens** [comment]
Please open a new issue if there is still a problem. Looks like this issue covers some different problems and a lot has changed.
